libraries/chain/backing_store/db_context_rocksdb.cpp

```diff
@@ -0,0 +1,749 @@
+
+      const auto old_payer = key_store.payer;
+
+      std::string event_id;
+      auto dm_logger = context.control.get_deep_mind_logger();
+      if (dm_logger != nullptr) {
+         event_id = db_context::table_event(table_store.contract, table_store.scope, table_store.table, name(key_store.primary));
+      }
+
+      payer_payload pp(*old_key_value.value);
+      update_db_usage( old_payer,  -((int64_t)pp.value_size + db_key_value_any_lookup::overhead), db_context::row_rem_trace(context.get_action_id(), event_id) );
+
+      if (dm_logger != nullptr) {
+         db_context::write_row_remove(*dm_logger, context.get_action_id(), table_store.contract, table_store.scope,
+                                      table_store.table, old_payer, name(key_store.primary), pp.value,
+                                      pp.value_size);
+      }
+
+      current_session.erase(old_key_value.full_key);
+      primary_lookup.remove_table_if_empty(old_key_value.full_key);
+
+      primary_iter_store.remove(itr); // don't use key_store anymore
+   }
+
+   int32_t db_context_rocksdb::db_get_i64(int32_t itr, char* value , size_t value_size) {
+      const auto& key_store = primary_iter_store.get(itr);
+      const auto& table_store = primary_iter_store.get_table(key_store);
+      const auto old_key_value = get_primary_key_value(table_store.contract, table_store.scope, table_store.table, key_store.primary);
+
+      EOS_ASSERT( old_key_value.value, db_rocksdb_invalid_operation_exception,
+                  "invariant failure in db_get_i64, iter store found to update but nothing in database");
+      payer_payload pp {*old_key_value.value};
+      const char* const actual_value = pp.value;
+      const size_t actual_size = pp.value_size;
+      if (value_size == 0) {
+         return actual_size;
+      }
+      const size_t copy_size = std::min<size_t>(value_size, actual_size);
+      memcpy( value, actual_value, copy_size );
+      return copy_size;
+   }
+
+   int32_t db_context_rocksdb::db_next_i64(int32_t itr, uint64_t& primary) {
+      if (itr < primary_iter_store.invalid_iterator()) return primary_iter_store.invalid_iterator(); // cannot increment past end iterator of table
+
+      const auto& key_store = primary_iter_store.get(itr);
+      const auto& table_store = primary_iter_store.get_table(key_store);
+      auto exact =
+            get_exact_iterator(table_store.contract, table_store.scope, table_store.table, key_store.primary);
+      EOS_ASSERT( exact.valid, db_rocksdb_invalid_operation_exception,
+                  "invariant failure in db_next_i64, iter store found to update but it does not exist in the database");
+      auto& session_iter = exact.itr;
+      auto& type_prefix = exact.type_prefix;
+      ++session_iter;
+      return find_and_store_primary_key(session_iter, key_store.table_ei, type_prefix,
+                                        key_store.table_ei, __func__, primary);
```

Comment:
__func__ is not used by `find_and_store_primary_key`

Reply:
Added it to the assert